cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
+from ..expressions.core import Comparison, Operator, BoolVal
 from ..expressions.globalconstraints import GlobalConstraint
-from ..expressions.utils import is_any_list, is_num, is_bool
+from ..expressions.utils import is_any_list, is_num, eval_comparison, is_bool
+
 from ..expressions.variables import _BoolVarImpl, boolvar, NegBoolView, _NumVarImpl
 
-def linearize_constraint(cpm_expr):
+def linearize_constraint(cpm_expr, supported={"sum","wsum"}, reified=False):
     """
     Transforms all constraints to a linear form.
     This function assumes all constraints are in 'flat normal form' with only boolean variables on the lhs of an implication.
     Only apply after 'cpmpy.transformations.flatten_model.flatten_constraint()' 'and only_bv_implies()'.
+
+    `AllDifferent` has a special linearization and is decomposed as such if not in `supported`.
+    Any other unsupported global constraint should be decomposed using `cpmpy.transformations.decompose_global.decompose_global()`
+
     """
 
     if is_any_list(cpm_expr):
-        lin_cons = [linearize_constraint(expr) for expr in cpm_expr]
+        lin_cons = [linearize_constraint(expr, supported=supported, reified=reified) for expr in cpm_expr]
         return [c for l in lin_cons for c in l]
 
+    # boolvar
     if isinstance(cpm_expr, _BoolVarImpl):
-        if isinstance(cpm_expr, NegBoolView):
-            return [cpm_expr._bv <= 0]
-        return [cpm_expr >= 1]
+        return [sum([cpm_expr]) >= 1]
 
-    if cpm_expr.name == "and":
-        if all(arg.is_bool() for arg in cpm_expr.args):
+    # Boolean operators
+    if isinstance(cpm_expr, Operator) and cpm_expr.is_bool():
+        # conjunction
+        if cpm_expr.name == "and":
             return [sum(cpm_expr.args) >= len(cpm_expr.args)]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
 
-    if cpm_expr.name == "or":
-        if all(arg.is_bool() for arg in cpm_expr.args):
+        # disjunction
+        elif cpm_expr.name == "or":
             return [sum(cpm_expr.args) >= 1]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
 
-    if cpm_expr.name == "->":
-        cond, sub_expr = cpm_expr.args
-        if not cond.is_bool() or not sub_expr.is_bool():
-            raise Exception(
-                f"Numeric constants or numeric variables not allowed as base constraint, cannot linearize {cpm_expr}")
-        lin_comps = linearize_constraint(sub_expr)
-        lin_exprs = []
-        for l_expr in lin_comps:
-            lhs, rhs = l_expr.args
-            if lhs.name == "mul" and _wsum_should(lhs):
-                # multiplication with var and constant, convert to wsum
-                lhs = Operator("wsum", _wsum_make(lhs))
+        # xor
+        elif cpm_expr.name == "xor" and len(cpm_expr.args) == 2:
+            return [sum(cpm_expr.args) >= 1]
```

Comment:
`return [sum(cpm_expr.args) == 1]` ?

Reply:
Note: for `xor` of arbitrary length, this works too if we allow an auxiliary variable, e.g., `xor(p,q,r,s,t)` becomes `p+q+r+s+t == 1+2*auxiliary`.